article/2021/threats-of-using-regular-expressions-in-javascript.md

```diff
@@ -44,96 +44,96 @@ gigiggt
+![运行结果](https://cdn-images-1.medium.com/max/2000/1*YUOV_B0E8SHaL_6ys3cDhQ.png)
 
-The reason for this issue was the overuse of the '+' operator inside the regex `/D[oD]?(\[[^\[\]]*\]|\s+)+MMMM?/` that caused this vulnerability. Luckily the issue was fixed on later versions after it was raised by [Snyk](https://snyk.io/) (a vulnerability tracking tool).
+`/D[oD]?(\[[^\[\]]*\]|\s+)+MMMM?/` 中 `+` 运算符的过度使用造成了这个漏洞。幸运的是，该问题由 [Snyk](https://snyk.io/)（一个漏洞追踪工具）提出后便在更高的版本中得到了修复。
 
-## How to Avoid Regex Vulnerabilities
+## 如何规避 RegEx 的漏洞
 
-### 1. Writing simple regex
+### 1. 编写简单的 RegEx
 
-Catastrophic backtracking occurs inside regexes that have at least three characters with two or more occurrences of the \*, +, } characters close to each other.
+当 RegEx 中包含至少 3 个字符，且包含至少两个彼此接近的 `*`、`+` 和 `}` 时，灾难性回溯会发生。
 
-So, if you can simplify your regex and avoid the above pattern, you can avoid catastrophic backtracking as well.
+所以，如果你能简化你的 RegEx 并避免使用以上的样式，那么你便能避免灾难性回溯。
 
-### 2. Use validation libraries
+### 2. 使用验证库
 
-For the most commonly used validation tasks, we can use third-party libraries like [validator.js](https://www.npmjs.com/package/validator) or [express-validator](https://www.npmjs.com/package/express-validator).
+对于常用的验证任务，我们可以使用第三方库，例如 [validator.js](https://www.npmjs.com/package/validator) 或 [express-validator](https://www.npmjs.com/package/express-validator)。
 
-We can depend on them as a large community backs them.
+我们可以依赖这些库，因为它们的背后有一个大型社区的支持。
 
-### 3. Use regex analyzers
+### 3. 使用 RegEx 分析器
 
-You can craft your own regex without any vulnerabilities using tools like [safe-regex](https://www.npmjs.com/package/safe-regex) and [rxxr2](https://www.cs.bham.ac.uk/~hxt/research/rxxr2/). They will check your regex for vulnerabilities and return its validity.
+你能通过使用 [safe-regex](https://www.npmjs.com/package/safe-regex)、[rxxr2](https://www.cs.bham.ac.uk/~hxt/research/rxxr2/) 等工具来编写无漏洞的 RegEx。它们将检查你的 RegEx 是否存在漏洞并返回其合法性。
 
 ```js
 var safe = require('safe-regex');
 
 var regex = /(g|i+)+t/;
-console.log(safe(regex)); //false
+console.log(safe(regex)); // false
 ```
 
-### 4. Avoid using Node’s default regex engine
+### 4. 避免使用 Node.js 默认的 RegEx 引擎
 
-Since Node’s default regex engine is vulnerable to ReDos attacks, we can avoid using it and switch to alternatives like google’s [re2](https://www.npmjs.com/package/re2) engine. It ensures that regexes are safe against ReDos attacks, and usage is almost similar to the default Node regex engine.
+由于 Node.js 默认的 RegEx 引擎容易受到 ReDoS 攻击，我们可以避免使用它，并以其他引擎作为替代，例如：Google 的 [re2](https://www.npmjs.com/package/re2) 引擎。它确保 RegEx 可以安全地抵御 ReDoS 攻击，用法也与 Node.js 默认的 RegEx 引擎相似。
 
 ```js
 var RE2 = require('re2');
 
 var re = new RE2(/(g|i+)+t/);
 var result = 'giiiiiiiiiiiiiiiiiiit'.search(re);
-console.log(result); //false
+console.log(result); // false
 ```
 
-This will evaluate to `false` because the regular expression is vulnerable to catastrophic backtracking.
```

Comment:
我本地运行了上方的这段代码，结果为 0，表示 search 函数运行了，也匹配成功了（将 'giiiiiiiiiiiiiiiiiiit' 改为 'giiiiiiiiiiiiiiiiiiiv' 结果不变）。这里作者的意思好像是因为这个 regex 可能发生 catastrophic backtracking 而不会运行（？？？）。

## Attention needed.

Reply:
我在本地也试过了，结果确实是 0，这没有问题，我认为原作者也确实应该是想要表达你说的意思。

>node-re2 can protect your Node.js application from ReDoS. node-re2 makes vulnerable regular expression patterns safe by evaluating them in RE2 instead of the built-in Node.js regex engine.

re2 的介绍里面也说了它确实可以抵御 ReDoS 攻击，所以原作者的介绍内容也没有问题

出现目前这种情况，我认为可能的原因有两个：
1. re2 抵御的行为和原作者描述不同，它可能会优化中间过程，最后返回结果，而不是完全不执行（没看过源码，这是我 YY 的，想知道答案就得去看看源码相关逻辑）
2. re2 的行为和原作者描述一致，只是作者给的这个例子可能还达不到 re2 认为的 ReDoS 攻击程度，所以就给执行了。作者在此处给出的例子可能只是表达一个形式，因为真的例子可能是一段很长的字符串，粘贴到文章中的话，影响文章排版。

我修改了示例代码，统计了它执行时间：
```js
var RE2 = require('re2');
var re = new RE2(/(g|i+)+t/);
console.time("共花费了");
var result = 'giiiiiiiiiiiiiiiiiiit'.search(re);
console.log(result); //false
console.timeEnd("共花费了");
```
![image](https://user-images.githubusercontent.com/32960305/132244024-4f05992c-73b0-43ac-a145-4342e3506801.png)
从这个执行时间来看，第二种可能性更大。

Reply:
我怎么感觉作者把案例 4 和案例 3 搞混掉了😂

Reply:
好像是耶 说到点了😃
@lsvih 大大，怎么处理 (๑•̌.•̑๑)ˀ̣ˀ̣

Reply:
@greycodee 你的意思是案例 4 的代码是从案例 3 复制过来的，但是作者忘了修改输出值？🤣 如果是这样的话，作者给的例子和他文章中表达的内容不太符合呢

Reply:
@KimYangOfCat 我觉得是 判定为 false 那个句子放错地方了

我试了长一点的字符串，多套几层 ()+ 也不 影响 re2 性能，我觉得你提出的第一种情况可能比较符合

（Mozilla 文档也说了 search 返回的是 number）

Reply:
这句话放错了吧。。

Reply:
> 这句话放错了吧。。

ok，我将这句话移上去~

Reply:
@lsvih ok done，可以 merge 了